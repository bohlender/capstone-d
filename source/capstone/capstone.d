/// Object-oriented wrapper of Capstone disassembly engine
module capstone.capstone;

import std.typecons: Tuple, BitFlags, Yes, Nullable;
import std.format: format;
import std.conv: to;
import std.array: array, appender;
import std.range: isInputRange, enumerate, front;
import std.algorithm: canFind;
import std.traits: EnumMembers;

import capstone.api;
import capstone.instruction;
import capstone.internal;
import capstone.error;

/** Encapsulates an instance of the Capstone dissassembly engine

This class encapsulates the core functionality of the Capstone disassembly engine, providing
access to runtime options for
$(UL
    $(LI changing the `Mode` of interpretation)
    $(LI changing the `Syntax` of the disassembly)
    $(LI choosing whether `Instruction`'s should be disassembled in detail, i.e. filling `Instruction.detail`)
    $(LI defining manual handling of broken instructions through the $(LINK2 http://www.capstone-engine.org/skipdata.html, SKIPDATA) mode of operation (optionally via a `Callback`))
)

Note that, since the architecture is chosen at runtime, this base class only provides access to the architecture-indepentent aspects,
e.g. disasm returns `Instruction`s instead of `X86Instruction`s.
However, if necessary, it can be casted to the architecture-specific variant, such as `CapstoneX86`.
*/
abstract class Capstone{
    package {
        alias Handle = size_t;
        Handle handle;    
        
        ModeFlags _mode;
        Syntax _syntax;
        bool _detail;
        bool _skipData;

        string mnemonic;
        Callback callback;

        string[int] customMnemonics;
    }
    const Arch arch; /// The architecture this Capstone instance is set up for

    /** Constructs an instance of the disassembly engine

    Params:
        arch = The architecture the engine will be created for
        modeFlags = A combination of flags to further specify how bytes will be interpreted, e.g. in little-endian.
    */
    private this(in Arch arch, in ModeFlags modeFlags){
        const libVer = versionOfLibrary;
        const bindVer = versionOfBindings;
        if(libVer != bindVer)
            throw new CapstoneException("API version mismatch between library (%s) and bindings (%s)".format(libVer, bindVer), ErrorCode.UnsupportedVersion);

        // Create Capstone engine instance
        this.arch = arch;
        this._mode = modeFlags;
        cs_open(arch, modeFlags.to!uint, &handle).checkErrno;
    }

    ~this(){
        if(handle)
           cs_close(&handle).checkErrno;
    }
    
    /// Gets the mode of interpretation
    @property auto mode() const {return _mode;}
    /// Sets the mode of interpretation
    @property void mode(in ModeFlags modeFlags){
        _mode = modeFlags;
        cs_option(handle, cs_opt_type.CS_OPT_MODE, modeFlags.to!uint).checkErrno;
    }

    /// Gets the disassembly syntax variant
    @property auto syntax() const {return _syntax;}
    /// Sets the disassembly syntax variant
    @property void syntax(in Syntax option){
        _syntax = option;
        cs_option(handle, cs_opt_type.CS_OPT_SYNTAX, option).checkErrno;
    }

    /// Indicates whether instructions will be disassembled in detail
    @property auto detail() const {return _detail;}
    /// Sets whether instructions will be disassembled in detail
    @property void detail(in bool enable){
        _detail = enable;
        auto option = enable ? cs_opt_value.CS_OPT_ON : cs_opt_value.CS_OPT_OFF;
        cs_option(handle, cs_opt_type.CS_OPT_DETAIL, option).checkErrno;
    }

    /// Indicates whether SKIPDATA mode of operation is in use
    @property auto skipData() const {return _skipData;}
    /// Sets whether to use SKIPDATA mode of operation
    @property void skipData(in bool enable){
        _skipData = enable;
        auto option = enable ? cs_opt_value.CS_OPT_ON : cs_opt_value.CS_OPT_OFF;
        cs_option(handle, cs_opt_type.CS_OPT_SKIPDATA, option).checkErrno;
    }

    /** Customises behaviour in SKIPDATA mode of operation
     
    By default, disassembling will stop when it encounters a broken instruction.
    Most of the time, the reason is that this is data mixed inside the input.

    When in SKIPDATA mode, some (unknown) amount of data until the next interpretable instruction will be skipped.
    Capstone considers the skipped data a special instruction with ID 0x00 and a `mnemonic` that defaults to `".byte"`.
    The operand string is a hex-code of the sequence of bytes it skipped.

    By default, for each iteration, Capstone skips 1 byte on X86 architecture, 2 bytes on Thumb mode on Arm
    architecture, and 4 bytes for the rest. The reason while Capstone skips 1 byte on X86 is that X86 puts no
    restriction on instruction alignment, but other architectures enforce some requirements on this aspect.

    To customise how many bytes to skip when encountering data, a `Callback` delegate can optonally be setup
    to return the corresponding number.

    Params:
        mnemonic = The mnemonic to use for representing skipped data
        callback = The optional callback to use for handling bytes that cannot be interpreted as an instruction.
    
    Example:
    ---
    // Custom data that can be referred to in a callback delegate
    struct CallbackData{
        int bytesToSkip;
    }
    auto myData = CallbackData(1);

    size_t myCallback(in ubyte[] code, size_t offset) {
        return myData.bytesToSkip++; // Always skip one more byte when encountering data
    }

    cs.skipData = true;                     // Enable skipdata mode
    cs.setupSkipdata("db", &myCallback);    // Use custom callback, and "db" as custom mnemonic for data
    ---
    */
    void setupSkipdata(in string mnemonic = ".byte", Callback callback = null){
        if(!mnemonic)
            throw new CapstoneException("Invalid mnemonic", ErrorCode.InvalidOption);
        this.mnemonic = mnemonic;
        this.callback = callback;
        
        auto setup = cs_opt_skipdata(this.mnemonic.ptr, this.callback ? &cCallback : null, &this.callback);
        cs_option(handle, cs_opt_type.CS_OPT_SKIPDATA_SETUP, cast(size_t)&setup).checkErrno;
    }

    /// Defines a custom mnemonic for a specified instruction id
    private void customMnemonic(in int id, in string mnem = null) {
        auto optMnem = cs_opt_mnem(id, null);
        if(mnem != null){
            const v = (customMnemonics[id] = mnem);
            optMnem.mnemonic = v.ptr;
        }else
            customMnemonics.remove(id);
        cs_option(handle, cs_opt_type.CS_OPT_MNEMONIC, cast(size_t)&optMnem).checkErrno;
    }

    /** Disassemble binary code, given the code buffer, start address and number of instructions to be decoded
    
    Params:
        code    = Buffer containing raw binary code to be disassembled
        address = Address of the first instruction in given raw code buffer
        count   = Number of instructions to be disassembled, or 0 to get all of them
    Returns: The successfully disassembled instructions

    Example:
    ---
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32)); // Initialise x86 32bit engine
    auto res = cs.disasm(CODE, 0x1000);               // Disassemble, offsetting addresses by 0x1000
    assert("%s %s".format(res[0].mnemonic, res[0].opStr) == "lea ecx, dword ptr [edx + esi + 8]");
    assert("%s %s".format(res[1].mnemonic, res[1].opStr) == "add eax, ebx");
    assert("%s %s".format(res[2].mnemonic, res[2].opStr) == "add esi, 0x1234");
    ---
    */
    abstract const(Instruction)[] disasm(in ubyte[] code, in ulong address, in size_t count = 0) const;

    /** Provides a range to iteratively disassemble binary code - one instruction at a time

    Fast API to disassemble binary code, given the code buffer and start address.
    Provides access to only one disassembled instruction at a time, resulting in a smaller memory footprint.
    Params:
        code    = Buffer containing raw binary code to be disassembled
        address = Address of the first instruction in given raw code buffer
    Returns: An input range over the disassembled instructions
    Example:
    ---
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32)); // Initialise x86 32bit engine
    auto range = cs.disasmIter(CODE, 0x1000);         // Disassemble one instruction at a time, offsetting addresses by 0x1000
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "lea ecx, dword ptr [edx + esi + 8]");
    range.popFront;
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "add eax, ebx");
    range.popFront;
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "add esi, 0x1234");
    range.popFront;
    assert(range.empty);
    ---
    */
    abstract InstructionRange disasmIter(in ubyte[] code, in ulong address) const;
}

// TODO: Try switching to InputRange!Instruction (more restrictive than isInputRange, though)
/// An input range that provides access to one disassembled `Instruction` at a time
abstract class InstructionRange {
    /// Retrieves element of the range
    Instruction front();
    /// True if range has no instructions, i.e. cannot be advanced anymore
    bool empty();
    /// Drops the front instruction and advances the range
    void popFront();
}
static assert(isInputRange!InstructionRange);

unittest{ // disasm
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = create(Arch.x86, ModeFlags(Mode.bit16));
    cs.mode = ModeFlags(Mode.bit32);

    auto res = cs.disasm(CODE, 0x1000);
    assert(res.length == 3); // With skipdata disabled, disassembling will halt when encountering data
    assert("%s %s".format(res[0].mnemonic, res[0].opStr) == "lea ecx, [edx + esi + 8]");
    assert("%s %s".format(res[1].mnemonic, res[1].opStr) == "add eax, ebx");
    assert("%s %s".format(res[2].mnemonic, res[2].opStr) == "add esi, 0x1234");
    cs.skipData = true;
    res = cs.disasm(CODE, 0x1000, 5);
    assert(res.length == 5);
}

/** Class template that encapsulates an architecture-specific instance of the Capstone dissassembly engine

Note that all architecture-specific instances, like `CapstoneX86`, instantiate and derive from this one.

Params:
    TInstructionId = The architecture-specific instruction identifier type
    TInstruction = The architecture-specific instruction type
*/
abstract class CapstoneImpl(TInstructionId, TInstruction) : Capstone {
    import capstone.range: InstructionImplRange;

    /** Creates an architecture-specific instance with a given mode of interpretation
    
    Params:
        arch = The architecture the engine will be created for
        modeFlags = The (initial) mode of interpretation, which can still be changed later on
    */
    package this(in Arch arch, in ModeFlags modeFlags){
        super(arch, modeFlags);
    }

    override TInstruction[] disasm(in ubyte[] code, in ulong address, in size_t count = 0) const {
        auto instrAppnd = appender!(TInstruction[]);
        foreach(i, instr; disasmIter(code, address).enumerate){
            instrAppnd.put(instr);
            if(i+1==count)
                break;
        }
        return instrAppnd.data;
    }

    override InstructionImplRange!TInstruction disasmIter(in ubyte[] code, in ulong address) const {
        return new InstructionImplRange!TInstruction(this, code, address);
    }

    /** Defines a custom mnemonic for a specified instruction id
    
    Example:
    ---
    enum X86_CODE32 = cast(ubyte[])"\x75\x01";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32));
    
    // Customize mnemonic JNE to JNZ
    cs.customMnemonic(X86InstructionId.jne, "jnz");
    
    foreach(instr; cs.disasm(X86_CODE32, 0x1000))
        writefln("%s\t%s", instr.mnemonic, instr.opStr);
    
    // Reset engine to use the default mnemonic of JNE
    cs.customMnemonic(X86InstructionId.jne);
    
    foreach(instr; cs.disasm(X86_CODE32, 0x1000))
        writefln("%s\t%s", instr.mnemonic, instr.opStr);
    ---
    */
    void customMnemonic(in TInstructionId id, in string mnem = null) {
        super.customMnemonic(id, mnem);
    }
}