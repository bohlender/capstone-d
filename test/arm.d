module test.arm;

import std.outbuffer;
import std.conv: to;

import capstone;
import test.utils;

enum ARM_CODE = cast(ubyte[])"\xED\xFF\xFF\xEB\x04\xe0\x2d\xe5\x00\x00\x00\x00\xe0\x83\x22\xe5\xf1\x02\x03\x0e\x00\x00\xa0\xe3\x02\x30\xc1\xe7\x00\x00\x53\xe3\x00\x02\x01\xf1\x05\x40\xd0\xe8\xf4\x80\x00\x00";
enum ARM_CODE2 = cast(ubyte[])"\xd1\xe8\x00\xf0\xf0\x24\x04\x07\x1f\x3c\xf2\xc0\x00\x00\x4f\xf0\x00\x01\x46\x6c";
enum THUMB_CODE = cast(ubyte[])"\x70\x47\x00\xf0\x10\xe8\xeb\x46\x83\xb0\xc9\x68\x1f\xb1\x30\xbf\xaf\xf3\x20\x84";
enum THUMB_CODE2 = cast(ubyte[])"\x4f\xf0\x00\x01\xbd\xe8\x00\x88\xd1\xe8\x00\xf0\x18\xbf\xad\xbf\xf3\xff\x0b\x0c\x86\xf3\x00\x89\x80\xf3\x00\x8c\x4f\xfa\x99\xf6\xd0\xff\xa2\x01";
enum THUMB_MCLASS = cast(ubyte[])"\xef\xf3\x02\x80";
enum ARMV8 = cast(ubyte[])"\xe0\x3b\xb2\xee\x42\x00\x01\xe1\x51\xf0\x7f\xf5";

enum platforms = [
	Platform(Arch.arm, Mode.arm, ARM_CODE, "ARM"),
	Platform(Arch.arm, Mode.armThumb, THUMB_CODE, "Thumb"),
	Platform(Arch.arm, Mode.armThumb, ARM_CODE2, "Thumb-mixed"),
	Platform(Arch.arm, Mode.armThumb, THUMB_CODE2, "Thumb-2 & register named with numbers", Syntax.noregname),
	Platform(Arch.arm, Mode.armThumb + Mode.armCortexM, THUMB_MCLASS, "Thumb-MClass"),
	Platform(Arch.arm, Mode.arm + Mode.armV8, ARMV8, "Arm-V8")
];

void writeDetail(ref OutBuffer buf, in Instruction!(Arch.arm) instr, in Capstone!(Arch.arm) cs){
	// TODO: No exception
	assert(!instr.detail.isNull);
	auto arm = instr.detail; //auto arm = instr.detail.archSpecific;
	
	if(arm.operands.length > 0)
		buf.writefln("\top_count: %d", arm.operands.length);
	foreach(i, operand; arm.operands){
		final switch(operand.type){
			case ArmOpType.INVALID:
				break;
			case ArmOpType.REG:
				buf.writefln("\t\toperands[%d].type: REG = %s", i, cs.regName(operand.reg));
				break;
			case ArmOpType.IMM:
				buf.writefln("\t\toperands[%d].type: IMM = 0x%x", i, operand.imm);
				break;
			case ArmOpType.FP:
				buf.writefln("\t\toperands[%d].type: FP = %f", i, operand.fp);
				break;
			case ArmOpType.MEM:
				buf.writefln("\t\toperands[%d].type: MEM", i);
				if (operand.mem.base != ArmRegister.INVALID)
					buf.writefln("\t\t\toperands[%d].mem.base: REG = %s", i, cs.regName(operand.mem.base));
				if (operand.mem.index != ArmRegister.INVALID)
					buf.writefln("\t\t\toperands[%d].mem.index: REG = %s", i, cs.regName(operand.mem.index));
				if (operand.mem.scale != 1)
					buf.writefln("\t\t\toperands[%d].mem.scale: %d", i, operand.mem.scale);
				if (operand.mem.disp != 0)
					buf.writefln("\t\t\toperands[%d].mem.disp: 0x%x", i, operand.mem.disp);
				break;
			case ArmOpType.PIMM:
				buf.writefln("\t\toperands[%d].type: P-IMM = %d", i, operand.imm);
				break;
			case ArmOpType.CIMM:
				buf.writefln("\t\toperands[%d].type: C-IMM = %d", i, operand.imm);
				break;
			case ArmOpType.SETEND:
				buf.writefln("\t\toperands[%d].type: SETEND = %s", i, operand.setend == ArmSetendType.BE? "be" : "le");
				break;
			case ArmOpType.SYSREG:
				buf.writefln("\t\toperands[%d].type: SYSREG = %d", i, operand.reg);
				break;
		}

		if (operand.shift.type != ArmShiftType.INVALID && operand.shift.value) {
			if (operand.shift.type < ArmShiftType.ASR_REG)
				// shift with constant value
				buf.writefln("\t\t\tShift: %d = %d", operand.shift.type, operand.shift.value);
			else
				// shift with register
				buf.writefln("\t\t\tShift: %d = %s", operand.shift.type, cs.regName(operand.shift.value.to!ArmRegister));
		}

		if (operand.vectorIndex != -1)
			buf.writefln("\t\toperands[%d].vector_index = %d", i, operand.vectorIndex);
		if (operand.subtracted)
			buf.writefln("\t\tSubtracted: True");
	}

	if (arm.cc != ArmCc.AL && arm.cc != ArmCc.INVALID)
		buf.writefln("\tCode condition: %d", arm.cc);
	if (arm.updateFlags)
		buf.writefln("\tUpdate-flags: True");
	if (arm.writeback)
		buf.writefln("\tWrite-back: True");
	if (arm.cpsMode)
		buf.writefln("\tCPSI-mode: %d", arm.cpsMode);
	if (arm.cpsFlag)
		buf.writefln("\tCPSI-flag: %d", arm.cpsFlag);
	if (arm.vectorData)
		buf.writefln("\tVector-data: %d", arm.vectorData);
	if (arm.vectorSize)
		buf.writefln("\tVector-size: %d", arm.vectorSize);
	if (arm.usermode)
		buf.writefln("\tUser-mode: True");
	if (arm.memBarrier)
		buf.writefln("\tMemory-barrier: %d", arm.memBarrier);
	buf.writefln("");
}

unittest{
	auto buf = new OutBuffer;
	static foreach(platform; platforms) {{
		auto cs = new Capstone!(platform.arch)(ModeFlags(platform.mode));
		cs.syntax = platform.syntax;
		cs.detail = true;
		
		auto res = cs.disasm(platform.code, 0x80001000);
		assert(res.length > 0);

		buf.writefln("****************");
		buf.writefln("Platform: %s", platform.comment);
		buf.writefln("Code:%s", platform.code.bytesToHex);
		buf.writefln("Disasm:");

		foreach(instr; res){
			buf.writefln("0x%x:\t%s\t%s", instr.address, instr.mnemonic, instr.opStr);
			buf.writeDetail(instr, cs);
		}
		buf.writefln("0x%x:", res[$-1].address + res[$-1].bytes.length);
		buf.writefln("");
	}}

	const expected = import("arm.expected");
	const actual = buf.toString;

	assert(expected == actual, expectationMismatch(expected, actual));
}