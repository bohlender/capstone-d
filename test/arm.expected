****************
Platform: ARM
Code:0x86 0x48 0x60 0xf4 0x4d 0x0f 0xe2 0xf4 0xed 0xff 0xff 0xeb 0x04 0xe0 0x2d 0xe5 0x00 0x00 0x00 0x00 0xe0 0x83 0x22 0xe5 0xf1 0x02 0x03 0x0e 0x00 0x00 0xa0 0xe3 0x02 0x30 0xc1 0xe7 0x00 0x00 0x53 0xe3 0x00 0x02 0x01 0xf1 0x05 0x40 0xd0 0xe8 0xf4 0x80 0x00 0x00 
Disasm:
0x80001000:	vld2.32	{d20, d21}, [r0], r6
	op_count: 4
		operands[0].type: REG = d20
		operands[0].access: WRITE
		operands[1].type: REG = d21
		operands[1].access: WRITE
		operands[2].type: MEM
			operands[2].mem.base: REG = r0
		operands[2].access: READ
		operands[3].type: REG = r6
		operands[3].access: READ
	Vector-size: 32
	Registers read: r0 r6
	Registers modified: d20 d21

0x80001004:	vld4.16	{d16[], d17[], d18[], d19[]}, [r2]!
	op_count: 5
		operands[0].type: REG = d16
		operands[0].access: WRITE
		operands[1].type: REG = d17
		operands[1].access: WRITE
		operands[2].type: REG = d18
		operands[2].access: WRITE
		operands[3].type: REG = d19
		operands[3].access: WRITE
		operands[4].type: MEM
			operands[4].mem.base: REG = r2
		operands[4].access: READ
	Write-back: True
	Vector-size: 16
	Registers read: r2
	Registers modified: d16 d17 d18 d19 r2

0x80001008:	bl	#0x80000fc4
	op_count: 1
		operands[0].type: IMM = 0x80000fc4
	Registers read: pc
	Registers modified: lr pc

0x8000100c:	str	lr, [sp, #-4]!
	op_count: 2
		operands[0].type: REG = lr
		operands[0].access: READ
		operands[1].type: MEM
			operands[1].mem.base: REG = sp
			operands[1].mem.disp: 0xfffffffc
		operands[1].access: WRITE
	Write-back: True
	Registers read: lr sp
	Registers modified: sp

0x80001010:	andeq	r0, r0, r0
	op_count: 3
		operands[0].type: REG = r0
		operands[0].access: WRITE
		operands[1].type: REG = r0
		operands[1].access: READ
		operands[2].type: REG = r0
		operands[2].access: READ
	Code condition: 1
	Registers read: r0
	Registers modified: r0

0x80001014:	str	r8, [r2, #-0x3e0]!
	op_count: 2
		operands[0].type: REG = r8
		operands[0].access: READ
		operands[1].type: MEM
			operands[1].mem.base: REG = r2
			operands[1].mem.disp: 0xfffffc20
		operands[1].access: WRITE
	Write-back: True
	Registers read: r8 r2
	Registers modified: r2

0x80001018:	mcreq	p2, #0, r0, c3, c1, #7
	op_count: 6
		operands[0].type: P-IMM = 2
		operands[1].type: IMM = 0x0
		operands[2].type: REG = r0
		operands[2].access: READ
		operands[3].type: C-IMM = 3
		operands[4].type: C-IMM = 1
		operands[5].type: IMM = 0x7
	Code condition: 1
	Registers read: r0

0x8000101c:	mov	r0, #0
	op_count: 2
		operands[0].type: REG = r0
		operands[0].access: WRITE
		operands[1].type: IMM = 0x0
	Registers modified: r0

0x80001020:	strb	r3, [r1, r2]
	op_count: 2
		operands[0].type: REG = r3
		operands[0].access: READ
		operands[1].type: MEM
			operands[1].mem.base: REG = r1
			operands[1].mem.index: REG = r2
		operands[1].access: WRITE
	Registers read: r3 r1 r2

0x80001024:	cmp	r3, #0
	op_count: 2
		operands[0].type: REG = r3
		operands[0].access: READ
		operands[1].type: IMM = 0x0
	Update-flags: True
	Registers read: r3
	Registers modified: cpsr

0x80001028:	setend	be
	op_count: 1
		operands[0].type: SETEND = be

0x8000102c:	ldm	r0, {r0, r2, lr} ^
	op_count: 4
		operands[0].type: REG = r0
		operands[0].access: READ
		operands[1].type: REG = r0
		operands[1].access: WRITE
		operands[2].type: REG = r2
		operands[2].access: WRITE
		operands[3].type: REG = lr
		operands[3].access: WRITE
	User-mode: True
	Registers read: r0
	Registers modified: r0 r2 lr

0x80001030:	strdeq	r8, sb, [r0], -r4
	op_count: 4
		operands[0].type: REG = r8
		operands[0].access: READ
		operands[1].type: REG = sb
		operands[1].access: READ
		operands[2].type: MEM
			operands[2].mem.base: REG = r0
		operands[2].access: READ
		operands[3].type: REG = r4
		operands[3].access: READ
		Subtracted: True
	Code condition: 1
	Write-back: True
	Registers read: r8 sb r0 r4
	Registers modified: r0

0x80001034:

****************
Platform: Thumb
Code:0x60 0xf9 0x1f 0x04 0xe0 0xf9 0x4f 0x07 0x70 0x47 0x00 0xf0 0x10 0xe8 0xeb 0x46 0x83 0xb0 0xc9 0x68 0x1f 0xb1 0x30 0xbf 0xaf 0xf3 0x20 0x84 0x52 0xf8 0x23 0xf0 
Disasm:
0x80001000:	vld3.8	{d16, d17, d18}, [r0:0x40]
	op_count: 4
		operands[0].type: REG = d16
		operands[0].access: WRITE
		operands[1].type: REG = d17
		operands[1].access: WRITE
		operands[2].type: REG = d18
		operands[2].access: WRITE
		operands[3].type: MEM
			operands[3].mem.base: REG = r0
			operands[3].mem.disp: 0x40
		operands[3].access: READ
	Vector-size: 8
	Registers read: r0
	Registers modified: d16 d17 d18

0x80001004:	vld4.16	{d16[1], d17[1], d18[1], d19[1]}, [r0]
	op_count: 5
		operands[0].type: REG = d16
		operands[0].neon_lane = 1
		operands[0].access: WRITE
		operands[1].type: REG = d17
		operands[1].neon_lane = 1
		operands[1].access: WRITE
		operands[2].type: REG = d18
		operands[2].neon_lane = 1
		operands[2].access: WRITE
		operands[3].type: REG = d19
		operands[3].neon_lane = 1
		operands[3].access: WRITE
		operands[4].type: MEM
			operands[4].mem.base: REG = r0
		operands[4].access: READ
	Vector-size: 16
	Registers read: r0
	Registers modified: d16 d17 d18 d19

0x80001008:	bx	lr
	op_count: 1
		operands[0].type: REG = lr
		operands[0].access: READ
	Registers read: lr
	Registers modified: pc

0x8000100a:	blx	#0x8000102c
	op_count: 1
		operands[0].type: IMM = 0x8000102c
	Registers read: pc
	Registers modified: lr pc

0x8000100e:	mov	fp, sp
	op_count: 2
		operands[0].type: REG = fp
		operands[0].access: WRITE
		operands[1].type: REG = sp
		operands[1].access: READ
	Registers read: sp
	Registers modified: fp

0x80001010:	sub	sp, #0xc
	op_count: 2
		operands[0].type: REG = sp
		operands[0].access: READ | WRITE
		operands[1].type: IMM = 0xc
	Registers read: sp
	Registers modified: sp

0x80001012:	ldr	r1, [r1, #0xc]
	op_count: 2
		operands[0].type: REG = r1
		operands[0].access: WRITE
		operands[1].type: MEM
			operands[1].mem.base: REG = r1
			operands[1].mem.disp: 0xc
		operands[1].access: READ
	Registers read: r1
	Registers modified: r1

0x80001014:	cbz	r7, #0x8000101e
	op_count: 2
		operands[0].type: REG = r7
		operands[0].access: READ
		operands[1].type: IMM = 0x8000101e
	Registers read: r7

0x80001016:	wfi	

0x80001018:	cpsie.w	f
	CPSI-mode: 2
	CPSI-flag: 1

0x8000101c:	ldr.w	pc, [r2, r3, lsl #2]
	op_count: 2
		operands[0].type: REG = pc
		operands[0].access: WRITE
		operands[1].type: MEM
			operands[1].mem.base: REG = r2
			operands[1].mem.index: REG = r3
		operands[1].access: READ
			Shift: 2 = 2
	Registers read: r2 r3
	Registers modified: pc

0x80001020:

****************
Platform: Thumb-mixed
Code:0xd1 0xe8 0x00 0xf0 0xf0 0x24 0x04 0x07 0x1f 0x3c 0xf2 0xc0 0x00 0x00 0x4f 0xf0 0x00 0x01 0x46 0x6c 
Disasm:
0x80001000:	tbb	[r1, r0]
	op_count: 1
		operands[0].type: MEM
			operands[0].mem.base: REG = r1
			operands[0].mem.index: REG = r0
		operands[0].access: READ
	Registers read: r1 r0

0x80001004:	movs	r4, #0xf0
	op_count: 2
		operands[0].type: REG = r4
		operands[0].access: WRITE
		operands[1].type: IMM = 0xf0
	Update-flags: True
	Registers modified: r4

0x80001006:	lsls	r4, r0, #0x1c
	op_count: 3
		operands[0].type: REG = r4
		operands[0].access: WRITE
		operands[1].type: REG = r0
		operands[1].access: READ
		operands[2].type: IMM = 0x1c
	Update-flags: True
	Registers read: r0
	Registers modified: r4

0x80001008:	subs	r4, #0x1f
	op_count: 2
		operands[0].type: REG = r4
		operands[0].access: READ | WRITE
		operands[1].type: IMM = 0x1f
	Update-flags: True
	Registers read: r4
	Registers modified: r4

0x8000100a:	stm	r0!, {r1, r4, r5, r6, r7}
	op_count: 6
		operands[0].type: REG = r0
		operands[0].access: READ | WRITE
		operands[1].type: REG = r1
		operands[1].access: READ
		operands[2].type: REG = r4
		operands[2].access: READ
		operands[3].type: REG = r5
		operands[3].access: READ
		operands[4].type: REG = r6
		operands[4].access: READ
		operands[5].type: REG = r7
		operands[5].access: READ
	Write-back: True
	Registers read: r0 r1 r4 r5 r6 r7
	Registers modified: r0

0x8000100c:	movs	r0, r0
	op_count: 2
		operands[0].type: REG = r0
		operands[0].access: WRITE
		operands[1].type: REG = r0
		operands[1].access: READ
	Update-flags: True
	Registers read: r0
	Registers modified: cpsr r0

0x8000100e:	mov.w	r1, #0
	op_count: 2
		operands[0].type: REG = r1
		operands[0].access: WRITE
		operands[1].type: IMM = 0x0
	Registers modified: r1

0x80001012:	ldr	r6, [r0, #0x44]
	op_count: 2
		operands[0].type: REG = r6
		operands[0].access: WRITE
		operands[1].type: MEM
			operands[1].mem.base: REG = r0
			operands[1].mem.disp: 0x44
		operands[1].access: READ
	Registers read: r0
	Registers modified: r6

0x80001014:

****************
Platform: Thumb-2 & register named with numbers
Code:0x4f 0xf0 0x00 0x01 0xbd 0xe8 0x00 0x88 0xd1 0xe8 0x00 0xf0 0x18 0xbf 0xad 0xbf 0xf3 0xff 0x0b 0x0c 0x86 0xf3 0x00 0x89 0x80 0xf3 0x00 0x8c 0x4f 0xfa 0x99 0xf6 0xd0 0xff 0xa2 0x01 
Disasm:
0x80001000:	mov.w	r1, #0
	op_count: 2
		operands[0].type: REG = r1
		operands[0].access: WRITE
		operands[1].type: IMM = 0x0
	Registers modified: r1

0x80001004:	pop.w	{r11, pc}
	op_count: 2
		operands[0].type: REG = r11
		operands[0].access: WRITE
		operands[1].type: REG = pc
		operands[1].access: WRITE
	Registers read: sp
	Registers modified: sp r11 pc

0x80001008:	tbb	[r1, r0]
	op_count: 1
		operands[0].type: MEM
			operands[0].mem.base: REG = r1
			operands[0].mem.index: REG = r0
		operands[0].access: READ
	Registers read: r1 r0

0x8000100c:	it	ne
	Code condition: 2
	Registers modified: itstate

0x8000100e:	iteet	ge
	Code condition: 11
	Registers modified: itstate

0x80001010:	vdupne.8	d16, d11[1]
	op_count: 2
		operands[0].type: REG = d16
		operands[0].access: WRITE
		operands[1].type: REG = d11
		operands[1].access: READ
		operands[1].vector_index = 1
	Code condition: 2
	Vector-size: 8
	Registers read: d11
	Registers modified: d16

0x80001014:	msr	cpsr_fc, r6
	op_count: 2
		operands[0].type: SYSREG = 144
		operands[1].type: REG = r6
		operands[1].access: READ
	Registers read: r6

0x80001018:	msr	apsr_nzcvqg, r0
	op_count: 2
		operands[0].type: SYSREG = 259
		operands[1].type: REG = r0
		operands[1].access: READ
	Registers read: r0

0x8000101c:	sxtb.w	r6, r9, ror #8
	op_count: 2
		operands[0].type: REG = r6
		operands[0].access: WRITE
		operands[1].type: REG = r9
		operands[1].access: READ
			Shift: 4 = 8
	Registers read: r9
	Registers modified: r6

0x80001020:	vaddw.u16	q8, q8, d18
	op_count: 3
		operands[0].type: REG = q8
		operands[0].access: WRITE
		operands[1].type: REG = q8
		operands[1].access: READ
		operands[2].type: REG = d18
		operands[2].access: READ
	Vector-data: 10
	Registers read: q8 d18
	Registers modified: q8

0x80001024:

****************
Platform: Thumb-MClass
Code:0xef 0xf3 0x02 0x80 
Disasm:
0x80001000:	mrs	r0, eapsr
	op_count: 2
		operands[0].type: REG = r0
		operands[0].access: WRITE
		operands[1].type: SYSREG = 264
	Registers modified: r0

0x80001004:

****************
Platform: Arm-V8
Code:0xe0 0x3b 0xb2 0xee 0x42 0x00 0x01 0xe1 0x51 0xf0 0x7f 0xf5 
Disasm:
0x80001000:	vcvtt.f64.f16	d3, s1
	op_count: 2
		operands[0].type: REG = d3
		operands[0].access: WRITE
		operands[1].type: REG = s1
		operands[1].access: READ
	Vector-data: 17
	Registers read: s1
	Registers modified: d3

0x80001004:	crc32b	r0, r1, r2
	op_count: 3
		operands[0].type: REG = r0
		operands[0].access: WRITE
		operands[1].type: REG = r1
		operands[1].access: READ
		operands[2].type: REG = r2
		operands[2].access: READ
	Registers read: r1 r2
	Registers modified: r0

0x80001008:	dmb	oshld
	Memory-barrier: 2

0x8000100c:

